src/db.rs

```diff
@@ -301,20 +302,47 @@ impl DbInner {
             .map(|wal_sst| wal_sst.id.unwrap_wal_id())
             .collect::<Vec<_>>();
         let mut last_sst_id = wal_id_last_compacted;
+
+        // load all the sst_handles in memory
+        let mut sst_handles = Vec::new();
```

Comment:
The SST handles are pretty small usually, but they do contain the first key byte string (up to 65k each). So I think it's safer and more memory efficient to lazily load the SST handles rather than pre-loading them all.

What I was thinking was pre-load four `SstIterator`s into a variable like `sst_iterators` (popping them off `wal_sst_list`. Then, while `sst_iterators`.pop isn't empty, you iterate over that SST to recover it. Finally, you pop off `wal_sst_list` and push a new `SstIterator` into the end of the `sst_iterators` vector.

The ordering is important since we want to play from oldest SST to newest, starting from last_compacted_wal_sst_id + 1.

Reply:
the problem with that is you will get a lifetime error if you try to instantiate an sst using `open_sst` inside the loop that consumes iterators because the sst instantiated inside the loop will have to be passed as a reference to new_spawn but as soon as the loop iteration ends sst would be dropped and new_spawn will have a dangling reference

consider the following pseudocode

```rust
// assume N iterators have been seeded into the vecdq

loop {

	// break if empty
	// iterators.pop_front()
    // exhaust iterator and build memtable

    let next_sst_id = remaining_wal_ssts.pop_front().expect("");
    let sst = self.table_store.open_sst(&SsTableId::Wal(sst_id)).await?;
    let id = match &sst.id {
                SsTableId::Wal(id) => *id,
                SsTableId::Compacted(_) => return Err(SlateDBError::InvalidDBState),
     };

	// sst doesn't live long enough
    let iter = SstIterator::new_spawn(&sst, self.table_store.clone(), 1, 256, true)
                        .await?;
   sst_iterators.push_back((iter, *sst_id));
       
}
```

Since `new_spawn` expects a reference to an sst handle and we want to `new_spawn` inside a loop we have to keep sst handles in memory outside the loop so that the references to them live long enough.

If you want I can push this version of the code so that you can play around with it locally and figure out if I am missing something.

Reply:
I think one reasonable way to do this would be to change `SstIterator` to store an `AsRef<SsTableHandle>` instead of `&SsTableHandle`:

```
pub(crate) struct SstIterator<'a, H: AsRef<SsTableHandle> = &'a SsTableHandle> {
    table: H,
    index: Arc<SsTableIndexOwned>,
    current_iter: Option<BlockIterator<Arc<Block>>>,
    from_key: Option<&'a [u8]>,
    next_block_idx_to_fetch: usize,
    fetch_tasks: VecDeque<FetchTask>,
    max_fetch_tasks: usize,
    blocks_to_fetch: usize,
    table_store: Arc<TableStore>,
    cache_blocks: bool,
}
```

You'd also need to add an impl for `AsRef` to `SsTableHandle`:
```
impl AsRef<SsTableHandle> for SsTableHandle {
    fn as_ref(&self) -> &SsTableHandle {
        self
    }
}
```

Then, you can create iterators that take an `Rc<SsTableHandle>`.

Reply:
that was a pretty neat trick, thank you. made the changes, it works well